clippy_lints/src/invalid_build_cfg.rs

```diff
@@ -0,0 +1,308 @@
+                                }
+                            } else {
+                                *has_unknown = true;
+                                return;
+                            }
+                        } else if [sym::windows, sym::unix].contains(&name) {
+                            out.push(CfgAst::Os(name));
+                        } else if [sym::any, sym::not, sym::all].contains(&name) {
+                            let mut sub_out = Vec::new();
+
+                            if let Some(next_token) = tokens.next()
+                                && let TokenTree::Delimited(_, _, _, stream) = next_token
+                            {
+                                parse_macro_args(stream, has_unknown, &mut sub_out);
+                            } else {
+                                *has_unknown = true;
+                            }
+                            if *has_unknown {
+                                // Could be set inside `parse_macro_args` too.
+                                return;
+                            }
+                            if name == sym::any {
+                                out.push(CfgAst::Any(sub_out));
+                            } else if name == sym::all {
+                                out.push(CfgAst::All(sub_out));
+                            } else if sub_out.len() > 1 {
+                                *has_unknown = true;
+                                return;
+                            } else if sub_out.len() == 1 {
+                                out.push(CfgAst::Not(Box::new(sub_out.pop().unwrap())));
+                            }
+                        } else {
+                            *has_unknown = true;
+                            return;
+                        }
+                    },
+                    TokenKind::Comma => {},
+                    _ => {
+                        *has_unknown = true;
+                        return;
+                    },
+                }
+            },
+            TokenTree::Delimited(_, _, _, stream) => parse_macro_args(stream, has_unknown, out),
+        }
+    }
+}
+
+fn get_invalid_cfg_attrs(attr: &MetaItem, spans: &mut Vec<Span>, has_feature: &mut bool) {
+    let Some(ident) = attr.ident() else { return };
+    if ident.name == sym::feature {
+        *has_feature = true;
+        return;
+    }
+    if attr.value_str().is_some() {
+        spans.push(attr.span);
+    } else if let Some(sub_attrs) = attr.meta_item_list() {
+        for sub_attr in sub_attrs {
+            if let Some(meta) = sub_attr.meta_item() {
+                get_invalid_cfg_attrs(meta, spans, has_feature);
+            }
+        }
+    } else {
+        spans.push(attr.span);
+    }
+}
+
+fn is_build_script(cx: &EarlyContext<'_>) -> bool {
+    cx.sess()
+        .opts
+        .crate_name
+        .as_ref()
+        .map_or(false, |crate_name| crate_name == "build_script_build")
+}
```

Comment:
It would also be good to check if we are actually being called by Cargo.
There is a nice function for that, [`rustc_session::utils::was_invoked_from_cargo`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_session/utils/fn.was_invoked_from_cargo.html).